packages/camera/camera_android/example/lib/camera_controller.dart

```diff
@@ -108,10 +109,10 @@ class CameraValue {
     bool? exposurePointSupported,
     bool? focusPointSupported,
     DeviceOrientation? deviceOrientation,
-    DeviceOrientation? lockedCaptureOrientation,
-    DeviceOrientation? recordingOrientation,
+    Optional<DeviceOrientation>? lockedCaptureOrientation,
+    Optional<DeviceOrientation>? recordingOrientation,
     bool? isPreviewPaused,
-    DeviceOrientation? previewPauseOrientation,
+    Optional<DeviceOrientation>? previewPauseOrientation,
```

Comment:
The combination of optional and nullable feels a bit dangerous to me.

Reply:
Fundamentally, to support the copyWith pattern with nullable values, you need to be able to express a nested nullable thing. It's unusual, but there isn't really a better pattern for it. The states we need to be able to represent are:
- I want to leave the old value unchanged
- I want to override the old value with a specific, non-null value
- I want to override the old value with null

Reply:
Just reading the code, it's unclear to me which one represent "leave the old value unchanged" and which one represent "override the old value with null". IMO using a more descriptive enum is clearer, despite of slightly more code.

Reply:
Please read other code review comments, as there has already been quite a bit of discussion on different approaches.

Reply:
@christopherfujino I read through the comments, but i'm not convinced that `Optional<X>?` is the best choice among alternatives. I think the benefit of clearer API outweights the drawback of slightly more coding (if it's more coding at all, since you are also copying the entire Optional class here).

Reply:
> IMO using a more descriptive enum is clearer

If the underlying value didn't happen to be an enum, how would you express this? The fact that we only have to do this (currently) with an enum value is a coincidence.

Reply:
> If the underlying value didn't happen to be an enum, how would you express this? 

Since dart does not support enum with associated value, we have to use a separate boolean flag (like `shouldLeaveOldValue` example you gave). 

> This seems fixable by renaming. E.g., OptionalReplacement with shouldLeaveOldValue (a bool) and newValue (a T) accessors.

Yes i think this approach is better than `Optional<T>?`. And since the underlying type is an enum, we can also combine the 2 parameters into 1 by defining a new enum. I think either is better than `Optional<T>?`.

Reply:
> Yes i think this approach is better than `Optional<T>?`.

To clarify, I'm suggesting something equivalent to (a stripped down) `Optional<T>` but called `OptionalReplacement<T>`.

>  And since the underlying type is an enum, we can also combine the 2 parameters into 1 by defining a new enum.

Again, that would only work for this one specific data type. We should not take a general problem that we are likely to need to solve in the future for another datatype, and intentionally solve it in a way that can't scale to those other types.

Reply:
> Again, that would only work for this one specific data type. We should not take a general problem that we are likely to need to solve in the future for another datatype, and intentionally solve it in a way that can't scale to those other types.

For a general solution, I think an `OptionalReplacement` with a `shouldLeaveOldValue` property makes sense - it combines the 2 parameters I talked about, and it works for all types. 

To make sure we are talking about the same thing, something like this right? 

```
class OptionalReplacement<T> {
  bool shouldLeaveOldValue
  T? replacement
}
```

Though I view it as "powered-up" (rather than "stripped-down") `Optional` since it adds a new capability to represent "leave old value".

Reply:
> ```
> class OptionalReplacement<T> {
>   bool shouldLeaveOldValue
>   T? replacement
> }
> ```

`T replacement`, because there's no reason it couldn't be used for both nullable and non-nullable types. (Arguably the API would be better if all the parameters used it.)

> Though I view it as "powered-up" (rather than "stripped-down") `Optional` since it adds a new capability to represent "leave old value".

Sorry, I forgot that this inverted the wrapping relative to the old semantics where it was the nullability of the wrapper that decided whether it was replacing or not. Despite the other issues, one advantage the old version had was that it was consistent about what a null parameter meant, which is an argument for keeping that version.

However, looking at the history a bit more here: the signature of the public `copyWith` method of a public type was changed without changing the major version of the plugin? That's bad (unless I'm misreading); changing the type of existing parameters is absolutely a breaking change. If that's indeed what happened, we should minimize the damage by replacing it with the copy of `Optional` with no changes because it minimize the chances that this break will break someone who hasn't yet updated to get the breaking change and been broken by it. If we want to actually redesign this API, it should be in a follow-up breaking change after the damage control fix goes out.

Reply:
> However, looking at the history a bit more here: the signature of the public `copyWith` method of a public type was changed without changing the major version of the plugin? That's bad (unless I'm misreading); changing the type of existing parameters is absolutely a breaking change. If that's indeed what happened, we should minimize the damage by replacing it with the copy of `Optional` with no changes because it minimize the chances that this break will break someone who hasn't yet updated to get the breaking change and been broken by it. If we want to actually redesign this API, it should be in a follow-up breaking change (with proper versioning) after the damage control fix goes out.

Agreed. I believe landing this PR as is would satisfy this (allowing users who were broken by the upgrade from 0.10.1 -> 0.10.2 to get essentially an API revert back to 0.10.1).